File: ~2023/03.programmers/2019_kakao_blind_recruitment/01.C++/03.cpp
#include <string>
#include <vector>
#include <iostream>
#include <set>
#include <algorithm>
using namespace std;

bool isPossible(vector<int> &vec,int now){
    for(int i=0;i<vec.size();i++)
        if((vec[i]&now)==vec[i])return false;
    return true;
}
int solution(vector<vector<string>> relation) {
    int n=relation.size();      // 행의 사이즈
    int m=relation[0].size();   // 열의 사이즈
    vector<int> ans;
    for(int i=1; i < (1<<m); i++){
        set<string> s;
        for(int j=0; j<n; j++){
            string now="";
            for(int k=0; k < m; k++){
                if(i & (1<<k)) now += relation[j][k];
            }
            s.insert(now);
        }
        if(s.size()==n&&isPossible(ans,i))ans.push_back(i);
    }
    return ans.size();
}
int main(){
    vector<vector<string>> rel ={
        {"100","ryan","music","2"},
        {"200","apeach","math","2"},
        {"300","tube","computer","3"},
        {"400","con","computer","4"},
        {"500","muzi","music","3"},
        {"600","apeach","music","2"}
    };
    
    cout<<solution(rel)<<endl;
    return 0;
}

File: ~2023/03.programmers/2019_kakao_blind_recruitment/02.Python/03.py
def isPossible(l, cur):
    for a in l:
        if (a&cur) == a: return False
    return True
def solution(relation):
    answer = []
    row = len(relation)
    col = len(relation[0])

    # 조합으로 나올 수 있는 모든 경우
    for i in range(1,1<<col):   
        s = set()
        for j in range(row):
            cur =""
            for k in range(col):
                #print("{0:b}".format(i).zfill(col)+" "+"{0:b}".format(1<<k).zfill(col))
                if i&(1<<k):
                    cur +=relation[j][k]
            s.add(cur)
        if len(s) == row and isPossible(answer,i):
            answer.append(i)
    return len(answer)

input_relation = [
    ["100","ryan","music","2"],
    ["200","apeach","math","2"],
    ["300","tube","computer","3"],
    ["400","con","computer","4"],
    ["500","muzi","music","3"],
    ["600","apeach","music","2"]
]
print(solution(input_relation))